### DISCUSSION TITLE
how to add damping term to a scalar wave equation

### DISCUSSION THREAD

**Discussion by aaelmeli** [idaholab/moose#20925] [Q&A General] [answered]
Hi

I have a scalar wave equation where I need to add a damping term to it. The equation is as follows:
![image](https://user-images.githubusercontent.com/88537645/166112893-3dff710f-978f-4efe-94b3-4f484ebec55a.png)
where `u`, is the scalar-valued function (the main variable). μ, η are the shear modulus and damping coefficient; respectively.
So, essentially it is diffusion term+damping term+ inertial term.

I could model the undamped system in the form: 
![image](https://user-images.githubusercontent.com/88537645/166112943-d64b8017-88dd-448e-86d7-97f15b6be9b9.png)
But now, I am not sure how to add the damping term!

Any ideas on how to model this?

**Comment by GiudGiud** [comment, marked as answer]
Hello

Is this because you want to model additional physical damping? Or because you want to numerically damp your equations? If the latter, you could try these objects:
https://mooseframework.inl.gov/syntax/Dampers/index.html

To get started on physical damping, I would do an integration by parts to remove the outer divergence term. Then you will be using the  gradSlnDot() function on the variable or preferably the coupledGradientDot(<variable_name>) from the Coupleable interface, to declare that term in the constructor of the object. Please let us know if you need help with that. 

Guillaume

**Reply by aaelmeli** [reply]
> Hello
> 
> Is this because you want to model additional physical damping? Or because you want to numerically damp your equations? If the latter, you could try these objects: https://mooseframework.inl.gov/syntax/Dampers/index.html
> 
> To get started on physical damping, I would do an integration by parts to remove the outer divergence term. Then you will be using the gradSlnDot() function on the variable or preferably the coupledGradientDot(<variable_name>) from the Coupleable interface, to declare that term in the constructor of the object. Please let us know if you need help with that.
> 
> Guillaume

Hi @GiudGiud , Yes, I am trying to add physical damping. Yes, please, I really appreciate your help.

**Reply by aaelmeli** [reply]
Hi @GiudGiud 
So, I am trying to start from where you suggested. Basically, the following is the `InputParameters ` and the `constructor `methods in `MatDiffusionBase.h`. So, I am adding the `zeta ` damping parameter as material properties in the parameter method, and `_grad_v_dot(coupledGradientDot("v")),` in the constructor. Sorry if this looks absolutely wrong, this is my first time implementing on moose, so please correct me. Also, when computing the residual, how best to make it work as the `ordinary diffusion kernel` and when having the `damping `term in addition to the `diffusion `term. Is the way I am implementing the `computeQpResidual() `correct?
```
template <typename T>
InputParameters
MatDiffusionBase<T>::validParams()
{
  InputParameters params = Kernel::validParams();
  params.addDeprecatedParam<MaterialPropertyName>(
      "D_name",
      "The name of the diffusivity",
      "This parameter has been renamed to 'diffusivity', which is more mnemonic and more conducive "
      "to passing a number literal");
  params.addParam<MaterialPropertyName>(
      "diffusivity", "D", "The diffusivity value or material property");
  params.addCoupledVar("args",
                       "Optional vector of arguments for the diffusivity. If provided and "
                       "diffusivity is a derivative parsed material, Jacobian contributions from "
                       "the diffusivity will be automatically computed");
  params.addCoupledVar("conc", "Deprecated! Use 'v' instead");
  params.addCoupledVar("v",
                       "Coupled concentration variable for kernel to operate on; if this "
                       "is not specified, the kernel's nonlinear variable will be used as "
                       "usual");
                       //Abdo from here
###   params.addParam<MaterialPropertyName>("zeta",
                                        0.0,
                                        "Name of material property or a constant real "
                                        "number defining the zeta parameter for the "
                                        "stiffness proportional damping.");                     
  return params;
}
```

```
template <typename T>
MatDiffusionBase<T>::MatDiffusionBase(const InputParameters & parameters)
  : DerivativeMaterialInterface<JvarMapKernelInterface<Kernel>>(parameters),
    _D(isParamValid("D_name") ? getMaterialProperty<T>("D_name")
                              : getMaterialProperty<T>("diffusivity")),
    _dDdc(getMaterialPropertyDerivative<T>(isParamValid("D_name") ? "D_name" : "diffusivity",
                                           _var.name())),
    _dDdarg(_coupled_moose_vars.size()),
    _is_coupled(isCoupled("v")),
    _v_var(_is_coupled ? coupled("v") : (isCoupled("conc") ? coupled("conc") : _var.number())),
    _grad_v(_is_coupled ? coupledGradient("v")
                        : (isCoupled("conc") ? coupledGradient("conc") : _grad_u)),
                        //Abdo from here
   ###  _grad_v_dot(coupledGradientDot("v")),
   ### _zeta(getMaterialProperty<Real>("zeta"))
{
  // deprecated variable parameter conc
  if (isCoupled("conc"))
    mooseDeprecated("In '", name(), "' the parameter 'conc' is deprecated, please use 'v' instead");

  // fetch derivatives
  for (unsigned int i = 0; i < _dDdarg.size(); ++i)
    _dDdarg[i] = &getMaterialPropertyDerivative<T>(
        isParamValid("D_name") ? "D_name" : "diffusivity", _coupled_moose_vars[i]->name());
}
```

```
template <typename T>
Real
MatDiffusionBase<T>::computeQpResidual()
{
    Real residual = 0.0;
  if (_static_initialization && _t == _dt)
  {
    residual +=
              _D[_qp] * _grad_v[_qp] * _grad_test[_i][_qp];
  }
   else if (_dt > 0)
  {
  residual +=
            _zeta[_qp] * _D[_qp] *_grad_test[_i][_qp] * _grad_v_dot[_qp];
  }
  return residual;
}
```

**Reply by GiudGiud** [reply]
The lines added to the constructor seem fine. 
Did you add the new class members in the header ?

```
  const MaterialProperty<Real> & _zeta;
  const VariableGradient & _grad_v_dot;
```

**Reply by GiudGiud** [reply]
You are trying to merge the damping and the diffusion kernels right? Like use one kernel for both terms in the equation?

**Reply by aaelmeli** [reply]
> const MaterialProperty<Real> & _zeta;
>   const VariableGradient & _grad_v_dot;

Thanks for noting this, I just added them.


> You are trying to merge the damping and the diffusion kernels right? Like use one kernel for both terms in the equation?

Exactly. Do you think this is better or you prefer having a separate kernel?

**Reply by GiudGiud** [reply]
Either way is OK. It's preferred to have a separate kernel if you foresee having to turn it on/off for debugging / prototyping for example. 

Does it compile now?

Where did you find this btw?
```
  if (_static_initialization && _t == _dt)
```
if start_time is below 0 that may never trigger for example.

**Reply by aaelmeli** [reply]
> Either way is OK. It's preferred to have a separate kernel if you foresee having to turn it on/off for debugging / prototyping for example.
> 
> Does it compile now?
> 
I am still trying to figure out how to implement the `computeQpJacobian()`, `computeQpOffDiagJacobian(unsigned int jvar)` and `computeQpCJacobian()`
> Where did you find this btw?
> 
> ```
>   if (_static_initialization && _t == _dt)
> ```

I borrowed it from `DynamicStressDivergenceTensors.C` in the definition of the `computeQpResidual()`
> 
> if start_time is below 0 that may never trigger for example.

I was thinking that way because I wanted something to switch on/off the damping term in the diffusion kernel.

**Reply by GiudGiud** [reply]
Once you have implemented that you can use the jacobian analyzer to make sure it's correct
https://mooseframework.inl.gov/help/development/analyze_jacobian.html

I'd use a boolean parameter to turn it on/off (if keeping in the same kernel).

**Reply by aaelmeli** [reply]
So, I had the following `computeQpJacobian()` modified so that it includes the damping term (utilizing a `boolean` parameter to switch it on/off). The jacobian term associated with the additional damping term should consists of 3 more terms to be added to the computeQpJacobian(), as shown. 

```
template <typename T>
Real
MatDiffusionBase<T>::computeQpJacobian()
{
  Real sum = _phi[_j][_qp] * _dDdc[_qp] * _grad_v[_qp] * _grad_test[_i][_qp];

  if (_add_stiffness_damping_term)
  {
    sum += _dzetadc[_qp] * _D[_qp] * _grad_test[_i][_qp] * _grad_v_dot[_qp]
        +  _zeta[_qp] * _dDdc[_qp] * _grad_test[_i][_qp] * _grad_v_dot[_qp]
        +  _zeta[_qp] * _D[_qp] * _grad_test[_i][_qp] *  _grad_phi[_j][_qp];
  }

  if (!_is_coupled)
    sum += computeQpCJacobian();

  return sum;
}
```
However, there is something that I do not understand, suppose we do not have any damping (i.e the original diffusion kernel), the jacobian term associated with the diffusion term would consist of the two terms, 
![image](https://user-images.githubusercontent.com/88537645/166298795-21dd21f7-b4c1-4512-b9db-bc7bfb8155f0.png)

one of them I can see is implemented in the `computeQpJacobian()`, and the other term, for some reason which I do not understand, is implemented in `computeQpCJacobian()`, which is apparently used when having a coupled variable. So, if you could help me fix this confusion in my understanding that would be really helpful.

Here is the other two computQp**Jacobian() methods for your reference.

```
template <typename T>
Real
MatDiffusionBase<T>::computeQpOffDiagJacobian(unsigned int jvar)
{
  // get the coupled variable jvar is referring to
  const unsigned int cvar = mapJvarToCvar(jvar);

  Real sum = (*_dDdarg[cvar])[_qp] * _phi[_j][_qp] * _grad_v[_qp] * _grad_test[_i][_qp];
  if (_v_var == jvar)
    sum += computeQpCJacobian();

  return sum;
}

template <typename T>
Real
MatDiffusionBase<T>::computeQpCJacobian()
{
  return _D[_qp] * _grad_phi[_j][_qp] * _grad_test[_i][_qp];
}
```

**Reply by GiudGiud** [reply]
I think the idea here is if the diffused variable is NOT the variable of the equation. 
Say you are solving for variable c and have a variable d in your equation
laplacian c + laplacian(d) + dc/dt = 0

then you use the coupled variable feature, and you have an off-diagonal term since the 'diagonal' is by definition terms for the same variable as the equation

**Reply by aaelmeli** [reply]
> Once you have implemented that you can use the jacobian analyzer to make sure it's correct https://mooseframework.inl.gov/help/development/analyze_jacobian.html

When using the jacobian analyzer, it produces tons of JSON files containing information like dofs, and I think node/element numbering! Do you know how to stop printing those files?

**Reply by GiudGiud** [reply]
I dont think we have that option. I would just manually run `rm *json` after running the script. It's really a punctual thing to do

**Reply by aaelmeli** [reply]
So, I am getting jacobian is 100% wrong! Apparently, I have made a fundamental mistake.
My kernel input looks as follows:

```
  [diffusion]
    type = MatDiffusion
    variable = p
    diffusivity = ShearModulus
    zeta= stiffness_damping
    v=p
    add_stiffness_damping_term=true    
  []
```
and the constructor is as follows:
```
template <typename T>
MatDiffusionBase<T>::MatDiffusionBase(const InputParameters & parameters)
  : DerivativeMaterialInterface<JvarMapKernelInterface<Kernel>>(parameters),
    _D(isParamValid("D_name") ? getMaterialProperty<T>("D_name")
                              : getMaterialProperty<T>("diffusivity")),
    _dDdc(getMaterialPropertyDerivative<T>(isParamValid("D_name") ? "D_name" : "diffusivity",
                                           _var.name())),
    _dDdarg(_coupled_moose_vars.size()),
    _is_coupled(isCoupled("v")),
    _v_var(_is_coupled ? coupled("v") : (isCoupled("conc") ? coupled("conc") : _var.number())),
    _grad_v(_is_coupled ? coupledGradient("v")
                        : (isCoupled("conc") ? coupledGradient("conc") : _grad_u)),
                        //Abdo from here
    _add_stiffness_damping_term(getParam<bool>("add_stiffness_damping_term")),                        
    _grad_v_dot(coupledGradientDot("v")),
    _zeta(getMaterialProperty<Real>("zeta")),
    _dzetadc(getMaterialPropertyDerivative<T>("zeta",_var.name())),
    _dzetadarg(_coupled_moose_vars.size())
{
  // deprecated variable parameter conc
  if (isCoupled("conc"))
    mooseDeprecated("In '", name(), "' the parameter 'conc' is deprecated, please use 'v' instead");

  // fetch derivatives
  for (unsigned int i = 0; i < _dDdarg.size(); ++i)
    _dDdarg[i] = &getMaterialPropertyDerivative<T>(
        isParamValid("D_name") ? "D_name" : "diffusivity", _coupled_moose_vars[i]->name());
  //Abdo: compute _dzetadarg
  for (unsigned int i = 0; i < _dzetadarg.size(); ++i)
    _dzetadarg[i] = &getMaterialPropertyDerivative<T>("zeta", _coupled_moose_vars[i]->name());        
}
```
So, if I understood it correctly, when specifying `v=p`, once an instance of this object is called, the constructor will compute the gradient of the time derivative of `v` (which is `p` here) i.e  `_grad_v_dot(coupledGradientDot("v"))`

Based on this, I implemented the `computeQpResidual `and the `computeQpJacobian`as follows:

```
template <typename T>
Real
MatDiffusionBase<T>::computeQpResidual()
{
    Real residual = 0.0;
    residual +=
              _D[_qp] * _grad_v[_qp] * _grad_test[_i][_qp];
  if (_add_stiffness_damping_term)
  {
    residual += _zeta[_qp] * _D[_qp] * _grad_test[_i][_qp] * _grad_v_dot[_qp];
  }
  return residual;
}
```
```
template <typename T>
Real
MatDiffusionBase<T>::computeQpJacobian()
{
  Real sum = _phi[_j][_qp] * _dDdc[_qp] * _grad_v[_qp] * _grad_test[_i][_qp];

   if (_add_stiffness_damping_term)
  {
    sum +=  _zeta[_qp] * _D[_qp] * _grad_test[_i][_qp] *  _grad_phi[_j][_qp];
  }

  if (!_is_coupled)
    sum += computeQpCJacobian();

  return sum;
}
```
```
template <typename T>
Real
MatDiffusionBase<T>::computeQpOffDiagJacobian(unsigned int jvar)
{
  // get the coupled variable jvar is referring to
  const unsigned int cvar = mapJvarToCvar(jvar);

  Real sum = (*_dDdarg[cvar])[_qp] * _phi[_j][_qp] * _grad_v[_qp] * _grad_test[_i][_qp];

  if (_v_var == jvar)
    sum += computeQpCJacobian();

  return sum;
}
```
```
template <typename T>
Real
MatDiffusionBase<T>::computeQpCJacobian()
{
   Real sum = _D[_qp] * _grad_phi[_j][_qp] * _grad_test[_i][_qp];

  return sum;
}
```
so, would you please have a look on this and correct my understanding?

**Reply by GiudGiud** [reply]
is the Jacobian right without the damping term?
Without you should be fairly close to the initial implementation

**Reply by GiudGiud** [reply]
also do you plan to have D AND zeta be functions of nonlinear variables? There's a little bit of code you dont need if not. 

I would also remove the references to "conc" from the code since it's deprecated. It ll be cleaner like this

**Reply by aaelmeli** [reply]
> is the Jacobian right without the damping term? Without you should be fairly close to the initial implementation

It looks correct when setting zeta=0. And the error increases linearly with increasing zeta! 

> ```
>   [diffusion]
>     type = MatDiffusion
>     variable = p
>     diffusivity = ShearModulus
>     zeta= stiffness_damping
>     v=p
>     add_stiffness_damping_term=true    
>   []
> ```

I am afraid of thins making the issue. Would using `p` as a `variable` and coupled variable `v` make trouble?

**Reply by aaelmeli** [reply]
> also do you plan to have D AND zeta be functions of nonlinear variables? There's a little bit of code you dont need if not.
> 

That is a good point. for now, we will consider zeta & diffusivity NOT functions of the nonlinear variable.

**Reply by GiudGiud** [reply]
so if there are no dependencies then you will be all set once you fix computeQpJacobian
Applying the derivative to the term in the integration by parts, I'd expect more terms with zeta

```
d(_phi[_j][_qp] * _d(zetaD)dc[_qp] * _grad_v[_qp] * _grad_test[_i][_qp] ) / dt
= d2 zeta dcdt  _phi[_j][_qp] * _d(D)dc[_qp] * _grad_v[_qp] * _grad_test[_i][_qp] + 
   dzeta dc  _phi[_j][_qp] * _d2(D)dcdt[_qp] * _grad_v[_qp] * _grad_test[_i][_qp] + 
   dzeta dc  _phi[_j][_qp] * _dDdc[_qp] * _grad_v_dot[_qp] * _grad_test[_i][_qp]
```
D and zeta are constants right?

If there are, you need to add the zeta dependence to computeQpOffDiagJacobian

**Reply by aaelmeli** [reply]
> so if there are no dependencies then you will be all set once you fix computeQpJacobian Applying the derivative to the term in the integration by parts, I'd expect more terms with zeta
> 
> ```
> d(_phi[_j][_qp] * _d(zetaD)dc[_qp] * _grad_v[_qp] * _grad_test[_i][_qp] ) / dt
> = d2 zeta dcdt  _phi[_j][_qp] * _d(D)dc[_qp] * _grad_v[_qp] * _grad_test[_i][_qp] + 
>    dzeta dc  _phi[_j][_qp] * _d2(D)dcdt[_qp] * _grad_v[_qp] * _grad_test[_i][_qp] + 
>    dzeta dc  _phi[_j][_qp] * _dDdc[_qp] * _grad_v_dot[_qp] * _grad_test[_i][_qp]
> ```
> 
> D and zeta are constants right?

Yes, they are constants. so if they are constants, I think the line of codes you wrote above would be equal to zero, right?

I could get the jacobian to work (based on the jacobian analyzer) with the following, where the missing thing was `_d_grad_v_dot_dv[_qp]`
Other than that, I have not modified any other jacobian methods.
```
template <typename T>
Real
MatDiffusionBase<T>::computeQpJacobian()
{
  Real sum = _phi[_j][_qp] * _dDdc[_qp] * _grad_v[_qp] * _grad_test[_i][_qp];
  if (_add_stiffness_damping_term)
  {
    sum += _zeta[_qp] * _grad_test[_i][_qp] * _grad_phi[_j][_qp] * _d_grad_v_dot_dv[_qp];
  }

  if (!_is_coupled)
    sum += computeQpCJacobian();

  return sum;
}
```

**Reply by GiudGiud** [reply]
One term is not zero. 
I dont know about d_grad_v_dot_dv to be honest. Glad it's correct though
